newbank/server/NewBank.java

```diff
@@ -1,70 +1,94 @@
 package newbank.server;
 
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.List;
-import java.util.Arrays;
 
 public class NewBank {
-  
+
   private static final NewBank bank = new NewBank();
-  private HashMap<String,Customer> customers;
-  
+  private HashMap<String, Customer> customers;
+  private boolean logOutRequested;
+
   private NewBank() {
     customers = new HashMap<>();
     addTestData();
+    logOutRequested = false;
   }
-  
+
   private void addTestData() {
+    // Password = 1
     Customer bhagy = new Customer();
     bhagy.addAccount(new Account("Main", 1000.0));
+    bhagy.assignPassword("c4ca4238a0b923820dcc509a6f75849b");
     customers.put("Bhagy", bhagy);
-    
+
+    // Password = 2
     Customer christina = new Customer();
     christina.addAccount(new Account("Savings", 1500.0));
+    christina.assignPassword("c81e728d9d4c2f636f067f89cc14862c");
     customers.put("Christina", christina);
-    
+
+    // Password = 3
     Customer john = new Customer();
     john.addAccount(new Account("Checking", 250.0));
+    john.assignPassword("eccbc87e4b5ce2fe28308fd9f2a7baf3");
     customers.put("John", john);
   }
-  
+
   public static NewBank getBank() {
     return bank;
   }
-  
+
   public synchronized CustomerID checkLogInDetails(String userName, String password) {
-    if(customers.containsKey(userName)) {
-      return new CustomerID(userName);
+    if (customers.containsKey(userName)) {
+      HashGenerator generator = new HashGenerator();
+      String hashedPassword = generator.generateHash(password);
+      String storedPassword = customers.get(userName).retrievePassword();
+      if (hashedPassword.equals(storedPassword)) {
+        return new CustomerID(userName);
+      }
     }
     return null;
   }
 
   // commands from the NewBank customer are processed in this method
   public synchronized String processRequest(CustomerID customer, String request) {
     if (customers.containsKey(customer.getKey())) {
-      
+
       List<String> tokens = Arrays.asList(request.split("\\s+"));
-      
-      if (tokens.size() > 0) {       
+
+      if (tokens.size() > 0) {
         switch (tokens.get(0)) {
-        case "SHOWMYACCOUNTS" : return showMyAccounts(customer);
-        case "NEWACCOUNT" : return addNewAccount(customer, tokens);
-        default : return "FAIL";
+          case "SHOWMYACCOUNTS":
+            return showMyAccounts(customer);
+          case "NEWACCOUNT":
+            return addNewAccount(customer, tokens);
+          case "LOGOUT":
+            logOutRequested = true;

```

Comment:
Hi Thomas I can't test this until later but wouldn't this log out every client connected to the server?  I think logOutRequested will need to be moved to a client-specific object.

Reply:
Hi Simon, my understanding is that the client handler creates a newbank object per client,  so this will only affect the client that has asked to logout, but I might be wrong, so please can you check this

Reply:
Hi Thomas I've just given it a try and it logs out all clients.  They've applied a singleton design pattern to the NewBank class so all client handlers reference the same static NewBank instance.

Reply:
Thanks Simon, I can see that now. I have added another commit and it only logs out one customer at a time. I did hope to put it in the customer class, however as it stands it was easier to place in the customerID class, let me know if that is ok?